kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+            cursor.execute('''
+                CREATE TABLE IF NOT EXISTS patterns (
+                    issue_id TEXT UNIQUE,
+                    issue_version INTEGER,
+                    pattern_object JSON
+                )
+            ''')
+            conn.commit()
+
+    def add_pattern(self, issue_id, issue_version, pattern_object):
+        """Add pattern object to DB"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('''
+                INSERT INTO patterns (issue_id, issue_version, pattern_object)
+                VALUES (?, ?, json(?))
+                ON CONFLICT(issue_id) DO UPDATE SET
+                issue_version=excluded.issue_version,
+                pattern_object=excluded.pattern_object
+            ''', (issue_id, issue_version, json.dumps(pattern_object)))
+            conn.commit()
+
+    def remove_pattern(self, issue_id):
+        """Remove pattern object from DB"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('DELETE FROM patterns WHERE issue_id = ?',
+                           (issue_id,))
+            conn.commit()
+
+    def get_all_patterns(self):
+        """Retrieve all patterns objects from DB"""
+        with sqlite3.connect(self.db_name) as conn:
+            cursor = conn.cursor()
+            cursor.execute('SELECT issue_id, issue_version, pattern_object '
+                           'FROM patterns')
+            while True:
+                row = cursor.fetchone()
+                if row is None:
+                    break
+                yield row
+
+    def update_patterns(self, issue):
+        """Update patterns for existing issue"""
+        if not issue.misc.get("pattern_object"):
+            self.remove_pattern(issue.id)
+            return
+
+        pattern_object = issue.misc.get("pattern_object")
+        if not PatternValidator().validate_pattern_object(pattern_object):
+            logger.error("Pattern object validation failed for issue id: %s",
+                         issue.id)
+            return
+        self.add_pattern(issue.id, issue.version, pattern_object)
+
+
+class PatternValidator:
+    """Pattern validator class"""
+    def __init__(self):
+        self.schema = copy.deepcopy(kcidb.io.SCHEMA.json)
+        self.remove_required_fields(self.schema)
+
+    def remove_required_fields(self, schema):
+        if "required" in schema:
+            del schema["required"]
+        if "properties" in schema:
+            for _, subschema in schema["properties"].items():
+                self.remove_required_fields(subschema)
+        if "items" in schema:
+            self.remove_required_fields(schema["items"])
+        if "$defs" in schema:
+            for _, subschema in schema["$defs"].items():
+                self.remove_required_fields(subschema)
+
+    def validate_pattern_object(self, pattern_object):
```

Comment:
"""Validate if pattern_object from the user was given as expected, as per KCIDB schema"""